Add tests and documentation for using `time_zone` in date range aggregation

cbuescher opened the pull request:
Date range aggregations used to be unable to use a `time_zone` parameter to e.g. be applied int date math roundings like in `now/d` (see #10130 as an example). After the aggregation refactoring, the `time_zone` parameter has been pulled up to ValuesSourceAggregatorBuilder and can now be used in date range aggregations as well. 

This change adds randomized time zone settings to the existing IT tests to verify that the `time_zone` parameter is honored when calculating the bucket boundaries. Also moving the DateRangeTests from module-groovy/messy back to core as DateRangeIT, sharing common script mocks with DateHistogramIT and adding documentation for the `time_zone` parameter in the date range aggregation docs.

Closes #10130

cbuescher committed:
Adding tests for `time_zone` parameter for date range aggregation

cbuescher committed:
Adding `time_zone` parameter to daterange-aggregation docs

nik9000 commented on buildSrc/src/main/resources/checkstyle_suppressions.xml:
```diff
@@ -1486,7 +1486,6 @@
   <suppress files="modules[/\\]lang-groovy[/\\]src[/\\]main[/\\]java[/\\]org[/\\]elasticsearch[/\\]script[/\\]groovy[/\\]GroovyScriptEngineService.java" checks="LineLength" />
   <suppress files="modules[/\\]lang-groovy[/\\]src[/\\]test[/\\]java[/\\]org[/\\]elasticsearch[/\\]messy[/\\]tests[/\\]BucketScriptTests.java" checks="LineLength" />
   <suppress files="modules[/\\]lang-groovy[/\\]src[/\\]test[/\\]java[/\\]org[/\\]elasticsearch[/\\]messy[/\\]tests[/\\]BulkTests.java" checks="LineLength" />
-  <suppress files="modules[/\\]lang-groovy[/\\]src[/\\]test[/\\]java[/\\]org[/\\]elasticsearch[/\\]messy[/\\]tests[/\\]DateRangeTests.java" checks="LineLength" />
```
Hurray!

nik9000 commented on core/src/test/java/org/elasticsearch/search/aggregations/bucket/DateRangeIT.java:
```diff
@@ -286,17 +287,22 @@ public void testSingleValueFieldWithStringDatesWithCustomFormat() throws Excepti
     }
 
     public void testSingleValueFieldWithDateMath() throws Exception {
+        int timeZoneOffset = randomIntBetween(-12, 12);
```
Can these have .5 and .75 (Nepal, why you have weird time zone?  :confused:) parts?

cbuescher replied:
I just found out you can go nuts and even have time zones with arbitrary minute offset like `-03.14` which might be the negative pi time zone, but as much as I like this (and Nepal) I'd prefer to leave it like this. I think those odd offsets don't add to the thing tested here and make the test harder to read.

nik9000 replied:
Nice! Fine by me leaving the test as is.

nik9000 commented on core/src/test/java/org/elasticsearch/search/aggregations/bucket/DateRangeIT.java:
```diff
@@ -286,17 +287,22 @@ public void testSingleValueFieldWithStringDatesWithCustomFormat() throws Excepti
     public void testSingleValueFieldWithDateMath() throws Exception {
+        int timeZoneOffset = randomIntBetween(-12, 12);
+        DateTimeZone timezone = DateTimeZone.forOffsetHours(timeZoneOffset);
+        String timeZoneSuffix = (timeZoneOffset == 0) ? "Z" : DateTime.now(timezone).toString("ZZ");
```
Could you leave an example of what this will produce so anyone scanning doesn't have to hunt down the "ZZ" format?

cbuescher replied:
sure

nik9000 commented on core/src/test/java/org/elasticsearch/search/aggregations/bucket/DateRangeTests.java:
```diff
@@ -56,6 +57,9 @@ protected DateRangeAggregatorBuilder createTestAggregatorBuilder() {
             factory.missing(randomIntBetween(0, 10));
         }
+        if (randomBoolean()) {
+            factory.timeZone(DateTimeZone.forOffsetHours(randomIntBetween(-23, 23)));
```
I wonder if we should grow a `randomTimeZone` method?

cbuescher replied:
I found LuceneTestCase#randomTimeZone() to be useful, will change that now.

cbuescher replied:
Not such a good idea, apparently Java TimeZone has some IDs which joda doesn't understand, so I'll stick with the zone IDs joda provides.

nik9000 commented on core/src/test/java/org/elasticsearch/search/aggregations/bucket/ScriptMocks.java:
```diff
@@ -0,0 +1,303 @@
+
+        @Override
+        public String description() {
+            return "Mock script engine for " + DateHistogramIT.class;
```
Probably worth putting an explanation in here.

nik9000 commented on core/src/test/java/org/elasticsearch/search/aggregations/bucket/ScriptMocks.java:
```diff
@@ -0,0 +1,303 @@
+    /**
+     * This mock script returns the field that is specified by name in the script body
+     */
+    public static class ExtractFieldScriptEngine implements ScriptEngineService {
```
Would a native script with the field name as a parameter be simpler?

nik9000 replied:
Talked with @cbuescher in a chat - since these are just copied from their old place they should probably just keep their implementation in this PR. Moving to test framework is still possible in this PR.

nik9000 replied:
Looks like there isn't an ExecutebleScript equivalent for search scripts anyway - ignore this.

nik9000 replied:
And I was wrong again! It looks like if your NativeScriptFactory returns something that extends AbstractSearchScript then you can indeed use native search scripts....

cbuescher replied:
can you elaborate? Not sure what that means and how this makes this simpler?

nik9000 replied:
Its cool. I don't think it is required for the PR. It would cut 50 lines or so - you implement NativeScriptFactory instead of ScriptEngineService and you implement AbstractSearchScript instead of LeafSearchLookup - just less code I guess.

nik9000 commented on core/src/test/java/org/elasticsearch/search/aggregations/bucket/ScriptMocks.java:
```diff
@@ -0,0 +1,303 @@
+    }
+
+    /**
+     * This mock script returns the field that is specified by name in the script body
```
Maybe explain that it is used in places where you want to make sure that scripts are valid but don't care about the specific script and this is the easiest way to do that.

nik9000 commented on core/src/test/java/org/elasticsearch/search/aggregations/bucket/ScriptMocks.java:
```diff
@@ -0,0 +1,303 @@
+    /**
+     * Mock plugin for the {@link ScriptMocks.ExtractFieldScriptEngine}
+     */
+    public static class ExtractFieldScriptPlugin extends Plugin {
```
I think it'd be nice to have this in :test:framework so others can use it.

cbuescher replied:
I'm not so sure about the usefulness of putting this in test:framework any more. There are several other variations of these mocks e.g. in AvgIT, SumIT, ValueCountIT, each having a slightly different mocking strategy. I'm actually thinking about moving them back to the respective IT tests, keep that stuff local. Generalizing all these mocks is an effort for a follow up PR I think.

nik9000 replied:
Fine by me.

nik9000 commented on core/src/test/java/org/elasticsearch/search/aggregations/bucket/ScriptMocks.java:
```diff
@@ -0,0 +1,303 @@
+    /**
+     * This mock script returns the field value and adds one month to the returned date
+     */
+    public static class FieldValueScriptEngine implements ScriptEngineService {
```
Maybe a name like "PlusOneMonth"

nik9000 commented on docs/reference/aggregations/bucket/daterange-aggregation.asciidoc:
```diff
@@ -111,3 +111,31 @@ Zone:: 'Z' outputs offset without a colon, 'ZZ' outputs the offset with a colon,
+  }
+--------------------------------------------------
+<1> This date will be converted to `2016-02-15T00:00:00.000+01:00`.
+<2> `now\d` will be rounded to the beginning of the day in the CET time zone.
```
Can you add a link to the list of supported time zones if it is handy? Or explain it somehow.

cbuescher replied:
sure

nik9000 replied:
Is `\` supposed to be `/`?

cbuescher replied:
yes, thanks for catching that, will change and squash a bit before merging.

cbuescher committed:
Simplify mock scripts

cbuescher commented:
@nik9000 thanks, I think I addressed most of your comments. I can also give extending NativeScriptFactory a shot, but need to see where that leads me and if it is simpler than the current mocks.

nik9000 commented:
> but need to see where that leads me and if it is simpler than the current mocks.

I think it'll be simpler but it isn't that important.

LGTM

cbuescher commented:
@nik9000 I switched the script mocks to native scripts and added Nepal-support for time zones ;-). I know you where already okay with this, but in case you want to take a last look before I merge.

nik9000 commented:
Left something small on the docs. LGTM otherwise

mj1856 commented on docs/reference/aggregations/bucket/daterange-aggregation.asciidoc:
```diff
@@ -111,3 +111,35 @@ Zone:: 'Z' outputs offset without a colon, 'ZZ' outputs the offset with a colon,
+Dates can be converted from another time zone to UTC by specifying the `time_zone` parameter.
+
+Time zones may either be specified as an ISO 8601 UTC offset (e.g. +01:00 or -08:00) or as one of
+the the http://joda-time.sourceforge.net/timezones.html[time zone ids] from the TZ database.
```
That URL isn't maintained very well.  A better link would be https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

Cheers!

clintongormley replied:
I've updated the link to use http://www.joda.org/joda-time/timezones.html instead, as that is the list that Joda actually ships with.

mj1856 replied:
Actually, the notes on that page say those are the time zones as of 2013d.  Joda current version 2.9.2 ships with tzdb 2016a - so there are a few time zones that have been added since then that aren't in the list.

Speaking of which, Wikipedia is at 2015g, I need to go update that to 2016b, I'll go do that now. :)

@jodastephen - Do you intend to keep the list of tzdb zones on joda.org updated? Or should a more neutral source like Wikipedia be prefered?  Also - are there any zones added that are not tzdb standard?  Thanks.

jodastephen replied:
The list at http://www.joda.org/joda-time/timezones.html tends to be very out of date. It is not regularly updated.

Joda-Time sticks to the IANA data, so there should not be additional zones.
